File: Amazon/Q1.cpp
//TLE using Python therefore also used C++ on the same question

// { Driver Code Starts
// Initial template for C++

#include <bits/stdc++.h>
using namespace std;

 // } Driver Code Ends
// User function Template for C++

class Solution {
  public:
    int maxProfit(int k, int n, int arr[]) {
        // code here
        int dp[k + 1][n];

        for(int i = 0; i <= k; i++) dp[i][0] = 0;
        for(int j = 0; j < n; j++) dp[0][j] = 0;
        
        for(int t = 1; t <= k; t++) {
        int mx = INT_MIN;
        for(int d = 1; d < n; d++) {
        
        mx = max(mx, dp[t - 1][d - 1] - arr[d - 1]);
        dp[t][d] = max(mx + arr[d], dp[t][d - 1]);
        }
        }
        
        return dp[k][n - 1];
    }
};

// { Driver Code Starts.
int main() {
    int t;
    cin >> t;
    while (t--) {
        int N, K;
        cin >> K;
        cin >> N;
        int A[N];
        for (int i = 0; i < N; i++) cin >> A[i];

        Solution ob;
        cout << ob.maxProfit(K, N, A) << endl;
    }
    return 0;
}  // } Driver Code Ends

File: Amazon/Q1.py
//TLE using Python therefore also used C++ on the same question
class Solution:
    def maxProfit(self, K, N, A):
        # code here
        profit = [[0 for i in range(K + 1)]for j in range(N)]
        
        for i in range(1, N):
            
            for j in range(1, K + 1):
                mx=0
                
                for l in range(i):
                    mx = max(mx,A[i]-A[l]+profit[l][j-1])
                
                profit[i][j] = max(profit[i - 1][j], mx)
        
        return profit[N-1][K]

#{ 
#  Driver Code Starts
#Initial Template for Python 3

if __name__ == '__main__':
    t = int(input())
    for _ in range(t):
        K = int(input())
        N = int(input())
        A = input().split()
        for i in range(N):
            A[i] = int(A[i])
        
        ob = Solution()
        print(ob.maxProfit(K, N, A))
# } Driver Code Ends
